json: Add top-level array decoding support

mrfuchs opened the pull request:
The library supports encoding JSON objects and arrays as well as
parsing JSON objects. Introduce a new function json_arr_parse() adding
support for parsing top-level JSON arrays.

Signed-off-by: Markus Fuchs <markus.fuchs@ch.sauter-bc.com>

mrfuchs committed:
json: Add top-level array decoding support

The library supports encoding JSON objects and arrays as well as
parsing JSON objects. Introduce a new function json_arr_parse() adding
support for parsing top-level JSON arrays.

Signed-off-by: Markus Fuchs <markus.fuchs@ch.sauter-bc.com>

Sir-Branch commented:
@mrfuchs Thanks for doing this LGTM, but is missing test cases. Give me a few days and I can write them and add to the PR

Sir-Branch commented:
@mrfuchs I went ahead and wrote the missing tests! Is the Allow edits by maintainers options enabled? I'm getting unauthorized when attempting to push the test commit to your branch

mrfuchs commented:
Yes, it's enabled. If it does not work just send me your code and I'll commit it.

Sir-Branch commented:
Looks like I'm limited to collaborator access, which probably doesn't allow commits into external feature branches. I went ahead and created a PR into your branch for visibility, feel free to adjust as needed or make a new commit

mrfuchs commented:
Thanks @Sir-Branch, I committed your changes.

Sir-Branch commented on tests/lib/json/src/main.c:
```diff
@@ -291,6 +329,77 @@ static void test_json_obj_arr_encoding(void)
+{
+	int ret;
+	struct obj_array obj_array_array_ts;
+	char encoded[] ="[{\"height\":168,\"name\":\"Simón Bolívar\"},"
```
Looks like I missed a space here after the =, sorry about that! Should be `char encoded[] = "[...` instead of `char encoded[] ="[...`

Sir-Branch commented on lib/os/json.c:
```diff
@@ -302,6 +302,23 @@ static int obj_init(struct json_obj *json, char *data, size_t len)
 
+static int arr_init(struct json_obj *json, char *data, size_t len)
+{
+	struct token token;
```
I'm not sure why `struct token token;` is tripping rule 5.7 when the same thing is done across multiple functions (obj_next, obj_init).
[Rule 5.7](https://docs.zephyrproject.org/latest/contribute/coding_guidelines/index.html): "A tag name shall be a unique identifier"

Sir-Branch replied:
@cfriedt Do you have any idea what might be causing this?

martinjaeger replied:
Maybe it's because the variable is named the same as the struct? Same variable name across different functions should really not be an issue.

In any case, rewriting with `struct token tok` should fix it.

Sir-Branch replied:
Hey Martin thanks for the help, I was guessing that might be it. I found it strange that it didn't trigger a warning for the obj_next and obj_init function which do the same thing (Maybe the rule check is only for new code?).
I'll go ahead and update all occurrences of `struct token token` to `struct token tok`, really appreciate the help!

martinjaeger replied:
Yep, it checks only the diff, AFAIK.

Sir-Branch committed:
tests: lib: json: Added tests array decoding, encoding and int limits

Added tests for:
- json_arr_parse
- json_arr_encode_buf
- integer limit testings

Signed-off-by: Ramiro Merello <rmerello@itba.edu.ar>

Sir-Branch committed:
json: Fix rule 5.7 violations (Tag name should be unique)

Fixed 5.7 rules errors due to:
- struct token token updated to struct token tok
- struct lexer lexer updated to struct lexer lex

Signed-off-by: Ramiro Merello <rmerello@itba.edu.ar>

Sir-Branch commented:
@mrfuchs I think I've fixed all the issues we were seeing, thanks for the patience. I created a PR Into your feature branch, hopefully this is enough to get it passing the cases

Sir-Branch commented:
@martinjaeger My review doesn't count as I've authored some of the commits, would you be able to review whenever you get a chance? Thanks again for all the help 😃

martinjaeger commented on tests/lib/json/src/main.c:
```diff
@@ -39,6 +39,12 @@ struct obj_array {
 
+struct test_int_limits {
+	int int_max;
+	int int_cero;
```
Spanish? I don't really mind as long as it's not a public API ;)

Sir-Branch replied:
Thanks so much for the quick review! Nice catch, the c and z will be the end of me 😄. Just in the internal testing, I'll fix it on my next PR :)
